application/common/models/Email.php

```diff
@@ -0,0 +1,284 @@
+            'to' => $this->to_address,
+            'subject' => $this->subject,
+            'status' => 'sent',
+        ]);
+        return 1;
+    }
+
+    /**
+     * Attempt to send email and on failure update attempts count and save (queue it)
+     * @throws Exception
+     * @return int number of sent messages
+     */
+    public function retry(): int
+    {
+        try {
+            return $this->send();
+        } catch (Exception $e) {
+            /*
+             * Send failed, attempt to queue
+             */
+            $this->attempts_count += 1;
+            $this->updated_at = time();
+
+            $log = [
+                'action' => 'retry sending email',
+                'to' => $this->to_address,
+                'subject' => $this->subject,
+                'attempts_count' => $this->attempts_count,
+                'last_attempt' => $this->updated_at,
+                'error' => $e->getMessage(),
+                'code' => $e->getCode()
+            ];
+            \Yii::error($log);
+
+            if (!$this->save()) {
+                \Yii::error([
+                    'action' => 'save email after failed retry failed',
+                    'status' => 'error',
+                    'error' => $this->getFirstErrors(),
+                ]);
+                throw new ServerErrorHttpException(
+                    'Unable to save email after failing to retry sending. Error: ' .
+                    print_r($this->getFirstErrors(), true),
+                    1500649788
+                );
+            }
+        }
+        return 0;
+    }
+
+    /**
+     * Builds a mailer object from $this and returns it
+     * @return \yii\mail\MessageInterface
+     */
+    public function getMessage()
+    {
+        $mailer = \Yii::$app->mailer->compose(
+            [
+                'html' => '@common/mail/html',
+                'text' => '@common/mail/text'
+            ],
+            [
+                'html' => $this->html_body,
+                'text' => $this->text_body
+            ]
+        );
+        $from = \Yii::$app->params['fromEmail'];
+        $name = \Yii::$app->params['fromName'];
+        if (empty($name)) {
+            $mailer->setFrom($from);
+        } else {
+            $mailer->setFrom([$from => $name]);
+        }
+        $mailer->setTo($this->to_address);
+        $mailer->setSubject($this->subject);
+
+        /*
+         * Conditionally set optional fields
+         */
+        $setMethods = [
+            'setCc' => $this->cc_address,
+            'setBcc' => $this->bcc_address,
+        ];
+        foreach ($setMethods as $method => $value) {
+            if ($value) {
+                $mailer->$method($value);
+            }
+        }
+
+        return $mailer;
+    }
+
+    /**
+     * Attempt to send messages from queue
+     * @throws Exception
+     */
+    public static function sendQueuedEmail()
+    {
+        $log = [
+            'action' => 'email/sendQueuedEmail',
```

Comment:
It might also be worth updating this, since it will presumably not be [the `actionSendQueuedEmail()` Controller Action](https://github.com/silinternational/email-service/blob/main/application/console/controllers/SendController.php#L13) calling this code now.

Then again, this function has that name, too, so the only part that doesn't quite fit any more is the `action` key in the log data. Feel free to proceed as you see fit.